### ISSUE TITLE
incompatible with mod_ruid2

### ISSUE THREAD

**Issue by GoogleCodeExporter** [pagespeed/mod_pagespeed#635]
```
I am running Apache 2.4 with mod_ruid2 and mod_pagespeed. I think the problem 
is mod_ruid2 because users run their own processes. So mod_pagespeed is trying 
to acces the folder(s) as a user instead of apache.

[code]
ls -ld /var/cache/mod_pagespeed
drwxr-xr-x 44 apache apache 4096 Mar  5 00:22 /var/cache/mod_pagespeed
[/code]

[code][Tue Mar 05 00:28:10.928752 2013] [pagespeed:error] [pid 8813] 
[mod_pagespeed 1.1.23.2-2258 @8813] Failed to mkdir 
/var/cache/mod_pagespeed/FtoZvGwwZ5ZVhZcMP-V9.outputlock: Permission denied
[Tue Mar 05 00:28:10.928780 2013] [pagespeed:error] [pid 8813] [mod_pagespeed 
1.1.23.2-2258 @8813] Failed to mkdir 
/var/cache/mod_pagespeed/fnFvXkOcs0MEl7oV4FzE.outputlock: Permission 
denied[/code]

Changing permission to 777 didn't help. How to solve this issue?
```

Original issue reported on code.google.com by `dy...@o2x.nl` on 4 Mar 2013 at 11:51

**Comment by jeffkaufman** [comment]
cPanel [documents](https://documentation.cpanel.net/display/EA/Apache+Module%3A+ModRuid2) that mod_ruid2 isn't compatible with `mod_cache` or the other normal apache caching modules for the same reason we're not: the permissions for cache files.  So we should probably just say we're not compatible with mod_ruid2.

@jmarantz did describe two workarounds above though:
- If you have mod_ruid2 configured on a per-VirtualHost basis, define
  a separate FileCachePath for each VirtualHost, or
- Use https://developers.google.com/speed/pagespeed/module/system#memcached
  instead of the File Cache

**Comment by jmarantz** [comment]
Wanted to capture this from ernexus:

"We are getting those errors very often (sometimes even few seconds apart for same users):
Failed to read cache clean timestamp /home/user/tmp/mod_pagespeed/!clean!time!. Doing an extra cache clean to be safe.

We do not get any other major errors, some "Fetch timed out" from time to time. My guess it could be due the overload of cache cleans."

**Comment by jeffkaufman** [comment]
Note additional details in https://github.com/pagespeed/mod_pagespeed/issues/1323, like `!clean!time!`

**Comment by ernexus** [comment]
Any news? Is there any plans to fix the issue in near future?

**Comment by jeffkaufman** [comment]
No plans to fix this, currently.  See my comment above https://github.com/pagespeed/mod_pagespeed/issues/635#issuecomment-171962790

The most reliable workaround would be to run PageSpeed on a separate server in front of your mod-ruid2 server, as a reverse proxy.  Something like nginx + ngx_pagespeed with proxy_pass.

**Comment by ernexus** [comment]
I that case we are going to compile our own Pagespeed version, which may solve the problem.

**Comment by jeffkaufman** [comment]
Cool; keep us posted!

**Comment by ernexus** [comment]
We managed to solve the file reading problem by compiling modified version of Pagespeed. We have changed the default permissions from 600 to 644 for !clean!time! and other temp files. It is not the best solution (would be better to fix owners), but it works in server with mod_ruid2.

**Comment by jmarantz** [comment]
ernexus -- we were discussing a code-change for lock-file permissions, but we think this will still have problems because the cache garbage-collector won't be able to delete files owned by another user.

In the other bug I suggested having a separate and disjoint FileCachePath per vhost.  Did you every try that?  I really think that's the way to go.

**Comment by ernexus** [comment]
Yes, we are currently using separate FileCachePath for each user. But the owners were still wrong, because of the mod_ruid2 (it takes the last owner of the Apache process and it could be anyone). We applied new folder creation rule for every user cache folder with ACL for 777, but the new files were still 600. So after the mod, files now 644. Because the folders have 777 i believe it can be purged. We are not getting any more errors after the modification.

It would be better to get proper owner, but we couldn't manage to get it working. Sorry i am not that familiar with Pagespeed code and programming language.

**Comment by anyandallart** [comment]
I'm running Centos 7 and Cpanel/WHM, and mod_ruid2 is enabled by default. I can get pagespeed to work if I disable mod_ruid2 and enable mod_suphp. I'm running a relatively small server so I just wanted to make sure there wasn't a big problem with that. I've read that mod_suphp is at "end of life," I just hope I'm not doing something that will affect me in the future. Any suggestions?
